File: android-sdk/tools/httpdns_request.py
#!/usr/bin/env python3
import argparse
import json
import os
import socket
import ssl
import sys
import time
import urllib.parse
from dataclasses import dataclass
from http.client import HTTPConnection, HTTPSConnection, HTTPResponse
from typing import Dict, List, Optional, Tuple

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


def env_or_default(name: str, default: str) -> str:
    value = os.getenv(name)
    return value if value is not None else default


def parse_bool(raw: str) -> bool:
    return raw.strip().lower() in {"1", "true", "yes", "on"}


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="HTTPDNS dispatch/resolve request tool")
    parser.add_argument("--mode", default=env_or_default("MODE", "dispatch"), choices=["dispatch", "resolve"])
    parser.add_argument("--scheme", default=env_or_default("SCHEME", "https"), choices=["https", "http"])

    parser.add_argument("--account-id", default=os.getenv("ACCOUNT_ID", ""))
    parser.add_argument("--aes-key", default=os.getenv("AES_KEY", ""))
    parser.add_argument(
        "--aes-key-mode",
        default=env_or_default("AES_KEY_MODE", "auto"),
        choices=["auto", "hex", "utf8"],
        help="AES key parse mode: auto|hex|utf8",
    )

    parser.add_argument("--bootstrap-host", default=env_or_default("BOOTSTRAP_HOST", "r.pp.fgnlo.com"))
    parser.add_argument("--bootstrap-ip", default=env_or_default("BOOTSTRAP_IP", ""))

    parser.add_argument("--resolve-host", default=env_or_default("RESOLVE_HOST", ""))
    parser.add_argument("--resolve-ip", default=env_or_default("RESOLVE_IP", ""))

    parser.add_argument("--region", default=env_or_default("REGION", "global"))
    parser.add_argument(
        "--dispatch-payload-format",
        default=env_or_default("DISPATCH_PAYLOAD_FORMAT", "proto"),
        choices=["proto", "json"],
        help="dispatch payload plain format",
    )
    parser.add_argument("--exp-seconds", type=int, default=int(env_or_default("EXP_SECONDS", "600")))
    parser.add_argument("--dn", default=env_or_default("DN", "www.example.com"))
    parser.add_argument("--dn-list", default=env_or_default("DN_LIST", ""), help="comma/space/semicolon separated hosts")
    parser.add_argument(
        "--resolve-batch-field",
        default=env_or_default("RESOLVE_BATCH_FIELD", "none"),
        choices=["none", "dn", "dns"],
        help="when dn-list is set: send one request with dn(csv) or dns(array) payload",
    )
    parser.add_argument("--q", default=env_or_default("Q", "4,6"))
    parser.add_argument("--cip", default=env_or_default("CIP", ""))
    parser.add_argument("--client-ip", default=env_or_default("CLIENT_IP", ""), help="alias of cip")
    parser.add_argument("--sdns-os", default=env_or_default("SDNS_OS", ""))

    parser.add_argument("--timeout-seconds", type=int, default=int(env_or_default("TIMEOUT_SECONDS", "10")))
    parser.add_argument("--disable-proxy", default=env_or_default("DISABLE_PROXY", "true"))
    parser.add_argument("--tls-verify", default=env_or_default("TLS_VERIFY", "true"))
    parser.add_argument("--ip-connect-sni", default=env_or_default("IP_CONNECT_SNI", "true"))
    parser.add_argument("--auto-fallback-http", default=env_or_default("AUTO_FALLBACK_HTTP", "false"))
    parser.add_argument("--show-http-only", default=env_or_default("SHOW_HTTP_ONLY", "false"))

    return parser.parse_args()


def parse_aes_key(raw_key: str) -> bytes:
    return parse_aes_key_with_mode(raw_key, "auto")


def parse_aes_key_with_mode(raw_key: str, mode: str) -> bytes:
    key_utf8 = raw_key.encode("utf-8")
    if mode == "utf8":
        if len(key_utf8) in {16, 24, 32}:
            return key_utf8
        raise ValueError("AES_KEY invalid for utf8 mode: expect utf8 length in {16,24,32}")

    if mode == "hex":
        if len(raw_key) in {32, 48, 64}:
            try:
                return bytes.fromhex(raw_key)
            except ValueError as error:
                raise ValueError("AES_KEY invalid hex") from error
        raise ValueError("AES_KEY invalid for hex mode: expect hex length in {32,48,64}")

    if len(key_utf8) in {16, 24, 32}:
        return key_utf8
    if len(raw_key) in {32, 48, 64}:
        try:
            return bytes.fromhex(raw_key)
        except ValueError:
            pass
    raise ValueError("AES_KEY invalid: expect utf8 length in {16,24,32} or hex length in {32,48,64}")


def encrypt_hex(aes_key: bytes, plaintext: bytes) -> str:
    iv = os.urandom(12)
    cipher_text = AESGCM(aes_key).encrypt(iv, plaintext, None)
    return (iv + cipher_text).hex()


def decrypt_hex(aes_key: bytes, hex_data: str) -> str:
    raw = bytes.fromhex(hex_data)
    if len(raw) <= 12:
        raise ValueError("invalid encrypted payload")
    iv = raw[:12]
    cipher_text = raw[12:]
    plaintext = AESGCM(aes_key).decrypt(iv, cipher_text, None)
    return plaintext.decode("utf-8")


def encode_varint(value: int) -> bytes:
    out = bytearray()
    while True:
        to_write = value & 0x7F
        value >>= 7
        if value:
            out.append(to_write | 0x80)
        else:
            out.append(to_write)
            return bytes(out)


def encode_length_delimited(field_number: int, raw: bytes) -> bytes:
    return bytes([(field_number << 3) | 2]) + encode_varint(len(raw)) + raw


def encode_varint_field(field_number: int, value: int) -> bytes:
    return bytes([(field_number << 3) | 0]) + encode_varint(value)


def build_dispatch_proto_plain(region: str, exp_ts: int) -> bytes:
    region_bytes = region.encode("utf-8")
    return encode_length_delimited(1, region_bytes) + encode_varint_field(3, exp_ts)


class SniHttpsConnection(HTTPSConnection):
    def __init__(self, connect_host: str, server_hostname: str, timeout: int, context: ssl.SSLContext, with_sni: bool):
        super().__init__(host=connect_host, port=443, timeout=timeout, context=context)
        self._server_hostname_override = server_hostname
        self._with_sni = with_sni

    def connect(self) -> None:
        address = (self.host, self.port)
        sock = socket.create_connection(address, self.timeout, self.source_address)
        if self._tunnel_host:
            self.sock = sock
            self._tunnel()
            sock = self.sock
        if self._with_sni:
            self.sock = self._context.wrap_socket(sock, server_hostname=self._server_hostname_override)
        else:
            self.sock = self._context.wrap_socket(sock)


@dataclass
class HttpResult:
    status: int
    reason: str
    headers: Dict[str, str]
    body: str


def make_request(
    scheme: str,
    host: str,
    path_with_query: str,
    connect_ip: str,
    timeout_seconds: int,
    disable_proxy: bool,
    tls_verify: bool,
    ip_connect_sni: bool,
) -> HttpResult:
    if disable_proxy:
        for key in ["http_proxy", "https_proxy", "HTTP_PROXY", "HTTPS_PROXY", "all_proxy", "ALL_PROXY"]:
            os.environ.pop(key, None)

    headers = {"Host": host} if connect_ip else {}
    target = connect_ip if connect_ip else host

    if scheme == "https":
        context = ssl.create_default_context() if tls_verify else ssl._create_unverified_context()
        context.check_hostname = tls_verify
        context.verify_mode = ssl.CERT_REQUIRED if tls_verify else ssl.CERT_NONE
        if connect_ip:
            conn = SniHttpsConnection(
                connect_host=target,
                server_hostname=host,
                timeout=timeout_seconds,
                context=context,
                with_sni=ip_connect_sni,
            )
        else:
            conn = HTTPSConnection(host=target, port=443, timeout=timeout_seconds, context=context)
    else:
        conn = HTTPConnection(host=target, port=80, timeout=timeout_seconds)

    try:
        conn.request("GET", path_with_query, headers=headers)
        response: HTTPResponse = conn.getresponse()
        raw = response.read()
        body = raw.decode("utf-8", errors="replace")
        return HttpResult(
            status=response.status,
            reason=response.reason,
            headers={k: v for k, v in response.getheaders()},
            body=body,
        )
    finally:
        conn.close()


def build_payload(args: argparse.Namespace, exp_ts: int) -> Tuple[bytes, str, str, str]:
    if args.mode == "dispatch":
        if args.dispatch_payload_format == "proto":
            payload = build_dispatch_proto_plain(region=args.region, exp_ts=exp_ts)
            payload_debug = f"proto_hex={payload.hex()}"
        else:
            payload_obj = {"region": args.region, "exp": exp_ts}
            payload = json.dumps(payload_obj, separators=(",", ":")).encode("utf-8")
            payload_debug = json.dumps(payload_obj, separators=(",", ":"))
        host = args.bootstrap_host
        connect_ip = args.bootstrap_ip
    else:
        if not args.resolve_host:
            raise ValueError("RESOLVE_HOST is required when MODE=resolve")
        hosts = parse_host_list(args.dn_list)
        if hosts and args.resolve_batch_field in {"dn", "dns"}:
            if args.resolve_batch_field == "dn":
                content = {"exp": exp_ts, "dn": ",".join(hosts), "q": args.q}
            else:
                content = {"exp": exp_ts, "dns": hosts, "q": args.q}
        else:
            target_dn = hosts[0] if hosts else args.dn
            content = {"exp": exp_ts, "dn": target_dn, "q": args.q}
        if args.cip:
            content["cip"] = args.cip
        if args.sdns_os:
            content["sdns-os"] = args.sdns_os
        payload_debug = json.dumps(content, separators=(",", ":"))
        payload = payload_debug.encode("utf-8")
        host = args.resolve_host
        connect_ip = args.resolve_ip
    return payload, host, connect_ip, payload_debug


def build_path(args: argparse.Namespace, account_id: str, enc: str) -> str:
    encoded_account = urllib.parse.quote(account_id, safe="")
    encoded_enc = urllib.parse.quote(enc, safe="")
    if args.mode == "dispatch":
        return f"/dnps-apis/v1/httpdns/endpoints?account_id={encoded_account}&enc={encoded_enc}"
    return f"/v1/d?id={encoded_account}&enc={encoded_enc}"


def print_request_context(args: argparse.Namespace, host: str, connect_ip: str, payload_debug: str, path: str) -> None:
    url = f"{args.scheme}://{host}{path}"
    print("=== Request Context ===")
    print(f"mode       : {args.mode}")
    print(f"scheme     : {args.scheme}")
    print(f"host       : {host}")
    print(f"connect_ip : {connect_ip if connect_ip else '<none>'}")
    print(f"cip        : {args.cip if args.cip else '<none>'}")
    print(f"payload    : {payload_debug}")
    print()
    print("=== HTTP Request ===")
    print(f"GET {path} HTTP/1.1")
    print(f"Host: {host}")
    print(f"URL : {url}")


def print_response(resp: HttpResult) -> None:
    print()
    print("=== Raw Response ===")
    print(f"status: {resp.status} {resp.reason}")
    print("headers:")
    for key, value in resp.headers.items():
        print(f"  {key}: {value}")
    print("body:")
    print(resp.body)


def parse_host_list(raw: str) -> List[str]:
    if not raw:
        return []
    normalized = raw.replace("x", ",").replace("X", ",")
    out: List[str] = []
    for token in normalized.replace(";", ",").replace("\n", ",").replace("\t", ",").replace(" ", ",").split(","):
        value = token.strip()
        if value:
            out.append(value)
    return out


def run_single(
    args: argparse.Namespace,
    aes_key: bytes,
    disable_proxy: bool,
    tls_verify: bool,
    ip_connect_sni: bool,
    auto_fallback_http: bool,
    show_http_only: bool,
) -> int:
    exp_ts = int(time.time()) + args.exp_seconds
    try:
        payload, host, connect_ip, payload_debug = build_payload(args, exp_ts)
    except ValueError as error:
        print(f"ERROR: {error}")
        return 1

    enc = encrypt_hex(aes_key, payload)
    path = build_path(args, args.account_id, enc)
    print_request_context(args, host, connect_ip, payload_debug, path)

    if show_http_only:
        return 0

    try:
        response = make_request(
            scheme=args.scheme,
            host=host,
            path_with_query=path,
            connect_ip=connect_ip,
            timeout_seconds=args.timeout_seconds,
            disable_proxy=disable_proxy,
            tls_verify=tls_verify,
            ip_connect_sni=ip_connect_sni,
        )
        print_response(response)
    except Exception as error:
        if args.scheme == "https" and auto_fallback_http:
            print()
            print(f"HTTPS request failed ({error}), retrying with HTTP...")
            try:
                response = make_request(
                    scheme="http",
                    host=host,
                    path_with_query=path,
                    connect_ip=connect_ip,
                    timeout_seconds=args.timeout_seconds,
                    disable_proxy=disable_proxy,
                    tls_verify=tls_verify,
                    ip_connect_sni=ip_connect_sni,
                )
                print_response(response)
            except Exception as second_error:
                print(f"ERROR: HTTP fallback failed: {second_error}")
                return 1
        else:
            print(f"ERROR: request failed: {error}")
            return 1

    print()
    print("=== Parsed Response ===")
    try:
        root = json.loads(response.body)
    except json.JSONDecodeError:
        print("Response is not JSON.")
        return 0

    data_hex = root.get("data", "")
    if not data_hex:
        print("No data field found in JSON response.")
        return 0

    print(f"data(hex): {data_hex}")
    print()
    print("=== Decrypted Data ===")
    try:
        print(decrypt_hex(aes_key, data_hex))
    except Exception as error:
        print(f"ERROR: decrypt failed: {error}")
        return 1

    return 0


def main() -> int:
    args = parse_args()
    if args.client_ip and not args.cip:
        args.cip = args.client_ip

    disable_proxy = parse_bool(args.disable_proxy)
    tls_verify = parse_bool(args.tls_verify)
    ip_connect_sni = parse_bool(args.ip_connect_sni)
    auto_fallback_http = parse_bool(args.auto_fallback_http)
    show_http_only = parse_bool(args.show_http_only)

    if not args.account_id or not args.aes_key:
        print("ERROR: ACCOUNT_ID and AES_KEY are required.")
        return 1

    try:
        aes_key = parse_aes_key_with_mode(args.aes_key, args.aes_key_mode)
    except ValueError as error:
        print(f"ERROR: {error}")
        return 1

    hosts = parse_host_list(args.dn_list)
    if args.mode == "resolve" and len(hosts) > 1 and args.resolve_batch_field == "none":
        exit_code = 0
        for index, host in enumerate(hosts, start=1):
            child_args = argparse.Namespace(**vars(args))
            child_args.dn = host
            child_args.dn_list = ""
            print()
            print(f"===== Resolve Host {index}/{len(hosts)}: {host} =====")
            code = run_single(
                child_args,
                aes_key,
                disable_proxy,
                tls_verify,
                ip_connect_sni,
                auto_fallback_http,
                show_http_only,
            )
            if code != 0:
                exit_code = code
        return exit_code

    return run_single(
        args,
        aes_key,
        disable_proxy,
        tls_verify,
        ip_connect_sni,
        auto_fallback_http,
        show_http_only,
    )


if __name__ == "__main__":
    sys.exit(main())

File: android-sdk/tools/httpdns_request.go
package main

import (
	"context"
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/tls"
	"encoding/hex"
	"encoding/json"
	"errors"
	"flag"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"
)

type config struct {
	Mode             string
	UseHTTPS         bool
	Scheme           string
	AccountID        string
	AESKey           string
	BootstrapHost    string
	BootstrapIP      string
	ResolveHost      string
	ResolveIP        string
	Region           string
	ExpSeconds       int
	DN               string
	Q                string
	CIP              string
	SdnsOS           string
	TimeoutSeconds   int
	DisableProxy     bool
	AutoFallbackHTTP bool
	ShowHTTPOnly     bool
}

func envOrDefault(name, defaultValue string) string {
	value := os.Getenv(name)
	if value == "" {
		return defaultValue
	}
	return value
}

func parseBool(raw string) bool {
	switch strings.ToLower(strings.TrimSpace(raw)) {
	case "1", "true", "yes", "on":
		return true
	default:
		return false
	}
}

func parseInt(raw string, defaultValue int) int {
	v, err := strconv.Atoi(strings.TrimSpace(raw))
	if err != nil {
		return defaultValue
	}
	return v
}

func parseArgs() config {
	defaultMode := envOrDefault("MODE", "dispatch")
	defaultUseHTTPS := parseBool(envOrDefault("HTTPS", "true"))
	defaultScheme := envOrDefault("SCHEME", "https")
	defaultAccountID := os.Getenv("ACCOUNT_ID")
	defaultAESKey := os.Getenv("AES_KEY")
	defaultBootstrapHost := envOrDefault("BOOTSTRAP_HOST", "r.pp.fgnlo.com")
	defaultBootstrapIP := envOrDefault("BOOTSTRAP_IP", "")
	defaultResolveHost := envOrDefault("RESOLVE_HOST", "")
	defaultResolveIP := envOrDefault("RESOLVE_IP", "")
	defaultRegion := envOrDefault("REGION", "global")
	defaultExpSeconds := parseInt(envOrDefault("EXP_SECONDS", "600"), 600)
	defaultDN := envOrDefault("DN", "www.example.com")
	defaultQ := envOrDefault("Q", "4,6")
	defaultCIP := envOrDefault("CIP", "")
	defaultSdnsOS := envOrDefault("SDNS_OS", "")
	defaultTimeout := parseInt(envOrDefault("TIMEOUT_SECONDS", "10"), 10)
	defaultDisableProxy := parseBool(envOrDefault("DISABLE_PROXY", "true"))
	defaultAutoFallback := parseBool(envOrDefault("AUTO_FALLBACK_HTTP", "false"))
	defaultShowHTTPOnly := parseBool(envOrDefault("SHOW_HTTP_ONLY", "false"))

	mode := flag.String("mode", defaultMode, "dispatch or resolve")
	useHTTPS := flag.Bool("https", defaultUseHTTPS, "use https (false means http)")
	scheme := flag.String("scheme", defaultScheme, "https or http")
	accountID := flag.String("account-id", defaultAccountID, "account id")
	aesKey := flag.String("aes-key", defaultAESKey, "aes key")
	bootstrapHost := flag.String("bootstrap-host", defaultBootstrapHost, "dispatch host")
	bootstrapIP := flag.String("bootstrap-ip", defaultBootstrapIP, "dispatch connect ip")
	resolveHost := flag.String("resolve-host", defaultResolveHost, "resolve host")
	resolveIP := flag.String("resolve-ip", defaultResolveIP, "resolve connect ip")
	region := flag.String("region", defaultRegion, "dispatch region")
	expSeconds := flag.Int("exp-seconds", defaultExpSeconds, "exp offset seconds")
	dn := flag.String("dn", defaultDN, "domain for resolve")
	q := flag.String("q", defaultQ, "query type, e.g. 4,6")
	cip := flag.String("cip", defaultCIP, "optional cip")
	sdnsOS := flag.String("sdns-os", defaultSdnsOS, "optional sdns-os")
	timeoutSeconds := flag.Int("timeout-seconds", defaultTimeout, "request timeout seconds")
	disableProxy := flag.Bool("disable-proxy", defaultDisableProxy, "disable proxy env")
	autoFallbackHTTP := flag.Bool("auto-fallback-http", defaultAutoFallback, "fallback to http when https request fails")
	showHTTPOnly := flag.Bool("show-http-only", defaultShowHTTPOnly, "print request and exit")

	flag.Parse()

	finalScheme := strings.ToLower(strings.TrimSpace(*scheme))
	if !*useHTTPS {
		finalScheme = "http"
	} else if finalScheme != "http" && finalScheme != "https" {
		finalScheme = "https"
	}

	return config{
		Mode:             *mode,
		UseHTTPS:         *useHTTPS,
		Scheme:           finalScheme,
		AccountID:        *accountID,
		AESKey:           *aesKey,
		BootstrapHost:    *bootstrapHost,
		BootstrapIP:      *bootstrapIP,
		ResolveHost:      *resolveHost,
		ResolveIP:        *resolveIP,
		Region:           *region,
		ExpSeconds:       *expSeconds,
		DN:               *dn,
		Q:                *q,
		CIP:              *cip,
		SdnsOS:           *sdnsOS,
		TimeoutSeconds:   *timeoutSeconds,
		DisableProxy:     *disableProxy,
		AutoFallbackHTTP: *autoFallbackHTTP,
		ShowHTTPOnly:     *showHTTPOnly,
	}
}

func parseAESKey(raw string) ([]byte, error) {
	utf8Bytes := []byte(raw)
	if len(utf8Bytes) == 16 {
		return utf8Bytes, nil
	}
	if len(raw) == 32 {
		decoded, err := hex.DecodeString(raw)
		if err == nil && len(decoded) == 16 {
			return decoded, nil
		}
	}
	return nil, errors.New("AES_KEY invalid: expect 16-byte plain text or 32-hex string")
}

func encryptHex(aesKey []byte, plaintext string) (string, error) {
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	iv := make([]byte, 12)
	if _, err := io.ReadFull(rand.Reader, iv); err != nil {
		return "", err
	}
	cipherText := gcm.Seal(nil, iv, []byte(plaintext), nil)
	out := append(iv, cipherText...)
	return hex.EncodeToString(out), nil
}

func decryptHex(aesKey []byte, hexData string) (string, error) {
	raw, err := hex.DecodeString(hexData)
	if err != nil {
		return "", err
	}
	if len(raw) <= 12 {
		return "", errors.New("invalid encrypted payload")
	}
	iv := raw[:12]
	cipherText := raw[12:]
	block, err := aes.NewCipher(aesKey)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}
	plain, err := gcm.Open(nil, iv, cipherText, nil)
	if err != nil {
		return "", err
	}
	return string(plain), nil
}

func buildPayload(cfg config, expTS int64) (string, string, string, error) {
	if cfg.Mode == "dispatch" {
		m := map[string]any{"region": cfg.Region, "exp": expTS}
		bytes, err := json.Marshal(m)
		if err != nil {
			return "", "", "", err
		}
		return string(bytes), cfg.BootstrapHost, cfg.BootstrapIP, nil
	}

	if cfg.Mode == "resolve" {
		if strings.TrimSpace(cfg.ResolveHost) == "" {
			return "", "", "", errors.New("RESOLVE_HOST is required when mode=resolve")
		}
		m := map[string]any{"exp": expTS, "dn": cfg.DN, "q": cfg.Q}
		if strings.TrimSpace(cfg.CIP) != "" {
			m["cip"] = cfg.CIP
		}
		if strings.TrimSpace(cfg.SdnsOS) != "" {
			m["sdns-os"] = cfg.SdnsOS
		}
		bytes, err := json.Marshal(m)
		if err != nil {
			return "", "", "", err
		}
		return string(bytes), cfg.ResolveHost, cfg.ResolveIP, nil
	}

	return "", "", "", errors.New("mode must be dispatch or resolve")
}

func buildPath(cfg config, accountID, enc string) string {
	encodedAccount := url.QueryEscape(accountID)
	encodedEnc := url.QueryEscape(enc)
	if cfg.Mode == "dispatch" {
		return fmt.Sprintf("/dnps-apis/v1/httpdns/endpoints?account_id=%s&enc=%s", encodedAccount, encodedEnc)
	}
	return fmt.Sprintf("/v1/d?id=%s&enc=%s", encodedAccount, encodedEnc)
}

func makeClient(cfg config, host, connectIP string) *http.Client {
	proxyFunc := http.ProxyFromEnvironment
	if cfg.DisableProxy {
		proxyFunc = nil
	}

	dialer := &net.Dialer{Timeout: time.Duration(cfg.TimeoutSeconds) * time.Second}

	transport := &http.Transport{
		Proxy: proxyFunc,
		DialContext: func(ctx context.Context, network, address string) (net.Conn, error) {
			target := address
			if strings.TrimSpace(connectIP) != "" {
				_, port, err := net.SplitHostPort(address)
				if err != nil {
					port = "443"
				}
				target = net.JoinHostPort(connectIP, port)
			}
			return dialer.DialContext(ctx, network, target)
		},
		TLSHandshakeTimeout: time.Duration(cfg.TimeoutSeconds) * time.Second,
		TLSClientConfig: &tls.Config{
			ServerName: host,
			MinVersion: tls.VersionTLS12,
		},
	}

	return &http.Client{
		Transport: transport,
		Timeout:   time.Duration(cfg.TimeoutSeconds) * time.Second,
	}
}

func doRequest(cfg config, host, pathWithQuery, connectIP string) (int, string, http.Header, string, error) {
	if strings.TrimSpace(connectIP) != "" {
		fmt.Printf("attempt endpoint: host=%s ip=%s\n", host, connectIP)
		return doSingleRequest(cfg, host, pathWithQuery, connectIP)
	}

	resolvedIPs, err := resolveHostIPs(host)
	if err != nil || len(resolvedIPs) == 0 {
		fmt.Printf("attempt endpoint: host=%s ip=<dns>\n", host)
		return doSingleRequest(cfg, host, pathWithQuery, "")
	}

	var lastErr error
	for _, ip := range resolvedIPs {
		fmt.Printf("attempt endpoint: host=%s ip=%s\n", host, ip)
		statusCode, statusText, headers, body, reqErr := doSingleRequest(cfg, host, pathWithQuery, ip)
		if reqErr == nil {
			return statusCode, statusText, headers, body, nil
		}
		fmt.Printf("attempt failed: ip=%s err=%v\n", ip, reqErr)
		lastErr = reqErr
	}

	if lastErr == nil {
		lastErr = errors.New("all endpoint attempts failed")
	}
	return 0, "", nil, "", lastErr
}

func doSingleRequest(cfg config, host, pathWithQuery, connectIP string) (int, string, http.Header, string, error) {
	baseURL := fmt.Sprintf("%s://%s%s", cfg.Scheme, host, pathWithQuery)
	requestURL, err := url.Parse(baseURL)
	if err != nil {
		return 0, "", nil, "", err
	}

	if strings.TrimSpace(connectIP) != "" {
		requestURL.Host = connectIP
	}

	req, err := http.NewRequest(http.MethodGet, requestURL.String(), nil)
	if err != nil {
		return 0, "", nil, "", err
	}
	if strings.TrimSpace(connectIP) != "" {
		req.Host = host
	}

	client := makeClient(cfg, host, connectIP)
	resp, err := client.Do(req)
	if err != nil {
		return 0, "", nil, "", err
	}
	defer resp.Body.Close()

	bodyBytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return resp.StatusCode, resp.Status, resp.Header, "", err
	}
	body := string(bodyBytes)
	return resp.StatusCode, resp.Status, resp.Header, body, nil
}

func resolveHostIPs(host string) ([]string, error) {
	ips, err := net.LookupIP(host)
	if err != nil {
		return nil, err
	}

	seen := make(map[string]struct{})
	out := make([]string, 0, len(ips))
	for _, ip := range ips {
		if ip == nil {
			continue
		}
		s := ip.String()
		if s == "" {
			continue
		}
		if _, exists := seen[s]; exists {
			continue
		}
		seen[s] = struct{}{}
		out = append(out, s)
	}
	return out, nil
}

func printRequestContext(cfg config, host, connectIP, payload, path string) {
	fmt.Println("=== Request Context ===")
	fmt.Printf("mode       : %s\n", cfg.Mode)
	fmt.Printf("scheme     : %s\n", cfg.Scheme)
	fmt.Printf("host       : %s\n", host)
	if strings.TrimSpace(connectIP) == "" {
		fmt.Println("connect_ip : <none>")
	} else {
		fmt.Printf("connect_ip : %s\n", connectIP)
	}
	fmt.Printf("payload    : %s\n", payload)
	fmt.Println()
	fmt.Println("=== HTTP Request ===")
	fmt.Printf("GET %s HTTP/1.1\n", path)
	fmt.Printf("Host: %s\n", host)
	fmt.Printf("URL : %s://%s%s\n", cfg.Scheme, host, path)
}

func printResponse(statusCode int, status string, headers http.Header, body string) {
	fmt.Println()
	fmt.Println("=== Raw Response ===")
	fmt.Printf("status: %d %s\n", statusCode, status)
	fmt.Println("headers:")
	for key, values := range headers {
		for _, value := range values {
			fmt.Printf("  %s: %s\n", key, value)
		}
	}
	fmt.Println("body:")
	fmt.Println(body)
}

func main() {
	cfg := parseArgs()

	if strings.TrimSpace(cfg.AccountID) == "" || strings.TrimSpace(cfg.AESKey) == "" {
		fmt.Println("ERROR: ACCOUNT_ID and AES_KEY are required")
		os.Exit(1)
	}

	aesKey, err := parseAESKey(cfg.AESKey)
	if err != nil {
		fmt.Printf("ERROR: %v\n", err)
		os.Exit(1)
	}

	expTS := time.Now().Unix() + int64(cfg.ExpSeconds)
	payload, host, connectIP, err := buildPayload(cfg, expTS)
	if err != nil {
		fmt.Printf("ERROR: %v\n", err)
		os.Exit(1)
	}

	enc, err := encryptHex(aesKey, payload)
	if err != nil {
		fmt.Printf("ERROR: encrypt failed: %v\n", err)
		os.Exit(1)
	}

	path := buildPath(cfg, cfg.AccountID, enc)
	printRequestContext(cfg, host, connectIP, payload, path)

	if cfg.ShowHTTPOnly {
		return
	}

	statusCode, statusText, headers, body, err := doRequest(cfg, host, path, connectIP)
	if err != nil && cfg.Scheme == "https" && cfg.AutoFallbackHTTP {
		fmt.Println()
		fmt.Printf("HTTPS request failed (%v), retrying with HTTP...\n", err)
		cfg.Scheme = "http"
		statusCode, statusText, headers, body, err = doRequest(cfg, host, path, connectIP)
	}
	if err != nil {
		fmt.Printf("ERROR: request failed: %v\n", err)
		os.Exit(1)
	}

	printResponse(statusCode, statusText, headers, body)

	fmt.Println()
	fmt.Println("=== Parsed Response ===")
	var root map[string]any
	if err := json.Unmarshal([]byte(body), &root); err != nil {
		fmt.Println("Response is not JSON.")
		return
	}

	dataAny, ok := root["data"]
	if !ok {
		fmt.Println("No data field found in JSON response.")
		return
	}

	dataHex, ok := dataAny.(string)
	if !ok || strings.TrimSpace(dataHex) == "" {
		fmt.Println("No data field found in JSON response.")
		return
	}

	fmt.Printf("data(hex): %s\n", dataHex)
	fmt.Println()
	fmt.Println("=== Decrypted Data ===")
	decrypted, err := decryptHex(aesKey, dataHex)
	if err != nil {
		fmt.Printf("ERROR: decrypt failed: %v\n", err)
		os.Exit(1)
	}
	fmt.Println(decrypted)
}
